### ISSUE TITLE
TODO

### ISSUE THREAD

**Issue by dylanaraps** [dylanaraps/neofetch#30]
#### Battery
- [ ] Multi-battery support for Windows and BSD.
- [ ] State support for multiple batteries.

#### Disk
- [ ] OpenBSD and NetBSD support.
- [ ] Print disk models

#### Testing

- [ ] Test neofetch on Qubes OS.

#### OS

- [ ] [ Puppy Linux ] Package count.

**Comment by kelaun** [comment]
Hi,

For the battery, the information is stored in "/sys/class/power_supply/BAT1/capacity" if you are using only one battery. Otherwise you have multiple files to query.

Hope this helps!

**Comment by dylanaraps** [comment]
Thanks!

What's the output of this?:

```sh
cat /sys/class/power_supply/BAT1/capacity
```

**Comment by kelaun** [comment]
It outputs the percentage of remaining battery, but only the number without the '%' (so like 89 for instance).

**Comment by dylanaraps** [comment]
Awesome!

Do you have multiple batteries? If so this command should print all of the battery info at once:

```sh
cat /sys/class/power_supply/BAT*/capacity
```

**Comment by kelaun** [comment]
I only have one battery but indeed it should output all the battery percentages.

**Comment by dylanaraps** [comment]
Is it possible to open the file in an editor? I want to know if it's just the battery value on line 1 like so:

```sh
1 80
2
3
```

**Comment by kelaun** [comment]
There is only one line in the file which contains the value.

**Comment by dylanaraps** [comment]
Awesome!

I'll get to work on this now.

**Comment by dylanaraps** [comment]
This is what the output will look like for a single battery:

```sh
Battery: 80%
```

How should we format the output for multiple batteries?

```sh
Battery: 1: 25%, 2: 50%, 3: 80%: 4. 20%

Battery: 25%, 50%, 80%, 20%
```

I'm not sure about how to go about this, prefixing each percentage with `BATX 20%` etc makes the line way too long.

**Comment by kelaun** [comment]
Can't you just create a new line for every battery that is present?
I would think that not that many laptops use more than one battery.

**Comment by tudurom** [comment]
It's not possible because of the config file structure. A solution would be enumerating them in a line.

**Comment by dylanaraps** [comment]
It **is** possible to multiline them, it'll just be some extra work.

**Comment by dylanaraps** [comment]
Here's the TODO for the battery function:

- Add support for all platforms
- Add an option to only print the first battery.
- Add an option to print all batteries on a single line.

**Comment by dylanaraps** [comment]
I'm going to release 1.1 when I've confirmed that the new battery function works on 
Mac OS X and Windows. If you guys are experiencing any bugs or have any suggestions
let me know and I'll squeeze them in before I do.

Thanks for all the help so far!

**Comment by dylanaraps** [comment]
@tudurom 
if the stdout stuff still works fine for you without any issues, I'll go ahead and release 1.1.

**Comment by tudurom** [comment]
I am doing my best trying to catch bugs!

**Comment by dylanaraps** [comment]
Thanks!

**Comment by mclado** [comment]
So, the IP option is still no-go? :thought_balloon:

**Comment by dylanaraps** [comment]
Oh, I've forgotten all about that!

What's the ideal output? The internel ip or external ip?

**Comment by mclado** [comment]
I had thoughts about external IP only but now when you also mentioned internal that would be great too. 

Also another though, I am gonna do some stuff on RPI those days so I am gonna test fetch there on GUI too. Will see how it will go if I asign his $os as Linux (using Raspbian). Meybe we can implement it too.

**Comment by dylanaraps** [comment]
Open a new issue about PI support, it shouldn't be too hard to add.

The first bit of info I'll need is the output of `uname` on the PI.

**Comment by dylanaraps** [comment]
I'll split IP into two functions, one that shows internal and one that shows external.

**Comment by mclado** [comment]
Great, tag me when you add IP support so I can test it!

Gonna do it when I try it out!

**Comment by dylanaraps** [comment]
Sure thing!

**Comment by dylanaraps** [comment]
@mclado, what's the output of this command?

```sh
echo $NET_IP
```

Also, does this program exist on your system?

```sh
ip
```

**Comment by tudurom** [comment]
What about BSD battery support? I think I'll install a VM.

**Comment by dylanaraps** [comment]
I'm getting a laptop in a couple of days so I was going to implement it then. 

Feel free to have a go in the meantime, you'll need to install NetBSD, OpenBSD
and FreeBSD as we'll probably have to do something different for each one.

Have a look through the functions in fetch and you'll see that each BSD stores
and formats the info differently.

Good luck, I'm here if you need any help.

**Comment by tudurom** [comment]
All 3 of them? It seems BSD isn't too standardized. Good luck, dear SSD.

**Comment by dylanaraps** [comment]
This command on `FreeBSD` gets the percentage:

```sh
acpiconf -i 0 | awk -F ': ' '/Remaining capacity/ {printf $2}
```

**Comment by dylanaraps** [comment]
This is how you get the perecentage on `OpenBSD`:

```sh
battery0=$(sysctl -n hw.sensors.acpibat0)

# Full amount
battery0full=$(sysctl -n hw.sensors.acpibat0.amphour0)
battery0full=${battery0full/ A*}

# Current amount
battery0now=$(sysctl -n hw.sensors.acpibat0.amphour3)
battery0now=${battery0now/ A*}

# Get the battery capacity in percentage
battery0=$(echo "100 * ${bat0now} / ${bat0full}" | bc)
battery0+="%"

```

Again, the `0` means battery number.

**Comment by dylanaraps** [comment]
For `NetBSD`, this should work:

```sh
apm | awk -F ' ' '/Battery remaining:/ {printf $2}'
```

I haven't tested any of these but they should get you started.

**Comment by dylanaraps** [comment]
I just did a fresh install of windows (Not in a VM) and the resolution command seems to work, so I'll push it to master.

```sh
wmic desktopmonitor get screenheight, screenwidth /value
```

**Comment by dylanaraps** [comment]
Leaving this here so I can get to it back on Linux.

```sh
width=$(wmic desktopmonitor get screenwidth /value 2>/dev/null)
width=${width/ScreenWidth'='/}
width=${width//[[:space:]]}

height=$(wmic desktopmonitor get screenheight /value 2>/dev/null)
height=${height/ScreenHeight'='/}
height=${height//[[:space:]]}

resolution="${width}x${height}"
```

**Comment by dylanaraps** [comment]
Windows resolution support is in master, I've only tested it with a single monitor
connected and it seems to work well.

**Comment by tudurom** [comment]
To get battery percentage on FreeBSD:

```bash
acpiconf -i 0 | awk -F ':\t' '/Remaining capacity/ {print $2}'
```

I will make a pr.

**Comment by dylanaraps** [comment]
Awesome!

**Comment by tudurom** [comment]
I have a problem with getting battery status on OpenBSD: there is no `hw.sensors.acpibat0.amphour0`. Instead, there's `hw.sensors.acpibat0.watthour0`.

**Comment by tudurom** [comment]
Ah, don't worry. I got this.

**Comment by tudurom** [comment]
Hey, OpenBSD support for battery is also done!

**Comment by mclado** [comment]
Good job!

**Comment by tudurom** [comment]
I learned a lot about git and BSD's doing this. Thanx!

**Comment by dylanaraps** [comment]
Once battery support is finished on all platforms I'm going to release `1.3`.

**Comment by dylanaraps** [comment]
I just went and bought a laptop, I should have it today or tommorrow.

**Comment by dylanaraps** [comment]
How do you guys feel about adding a new function to show logged in users?

```sh
# Show currently logged in users
getusers () {
    users="$(who | awk '!seen[$1]++ {printf $1 ", "}')"
    users=${s%,*}
}

# output
Users: dyl, root, jeff
```

**Comment by tudurom** [comment]
You may add it. Maybe there is someone in this world who wants to see the
logged in users on the lemonbar.

Also, NetBSD it hell (get it?). I warned you.

**Comment by dylanaraps** [comment]
This page says that you need to rebuild the BSD kernel to get the `apm` command
which prints the battery usage.

http://www.netbsd.org/docs/power-mgmt/

> First off, you need to compile a kernel with the APM driver enabled. Add (or uncomment) the following line in your kernel configuration file:
> `apm0	at mainbus0	# Advanced power management`

> Once you have a kernel with the APM kernel driver built, run the apm daemon (apmd) at boot time by setting the apmd variable /etc/rc.conf to YES.

> Once you've done both of these steps, you can use the apm utility to interface with the APM system. For example:

> ```sh
> % apm
> Battery charge state: high
> Battery remaining: 98 percent
> A/C adapter state: not connected
> Power management enabled
> ```

I'm still trying to see if there's any other way of getting the info.

**Comment by tudurom** [comment]
If it's required, then who wants battery output on NetBSD? Let's leave it for another release if there's demand. OpenBSD is made to run on all kinds of devices, not desktop computers.

**Comment by mclado** [comment]
Congratz on buying laptop, what have you bought?

**Comment by dylanaraps** [comment]
I got a Lenovo Yoga 900.

**Comment by mclado** [comment]
Well, looks nice. Hope it serve you well!

**Comment by dylanaraps** [comment]
Thanks! :)

**Comment by tudurom** [comment]
I have an idea: an option to opt for smaller distro ascii arts, like in [ufetch](https://github.com/jschx/ufetch).

**Comment by dylanaraps** [comment]
Good idea but how would the user swap between them?

We could add a new option: `ascii_distro_size`, how does that sound?

**Comment by tudurom** [comment]
That was my ideea, too!

**Comment by dylanaraps** [comment]
@tudurom, I've merged it.

**Comment by tudurom** [comment]
And now we shall add more mini ascii logos for the other distros. I'm looking for an easy method for doing this. My idea is to use [jp2a](https://github.com/cslarsen/jp2a) and then add the colors manually.

**Comment by dylanaraps** [comment]
Yup, colors have to be done manually regardless of what method we use to create ascii.

**Comment by dylanaraps** [comment]
I'm a little busy today and tomorrow, then I'm free. Sorry guys.

**Comment by mclado** [comment]
Any new ideas? :laughing:

**Comment by dylanaraps** [comment]
Give me some! :)

**Comment by tudurom** [comment]
Custom date format?

**Comment by dylanaraps** [comment]
The user can easily add that to their printinfo function no?

```sh
prin "Date: $(date 'flags here')"
```

**Comment by tudurom** [comment]
Yes. It would be nice if the user could specify a format for the date
returned by the birthday function.

**Comment by dylanaraps** [comment]
Ah

I'll add that now, good idea.

**Comment by dylanaraps** [comment]
Done.

**Comment by dylanaraps** [comment]
Any other ideas?

**Comment by dawidd6** [comment]
Could you add support for Alpine Linux?

http://www.alpinelinux.org

**Comment by dylanaraps** [comment]
Downloading the iso now.

**Comment by tudurom** [comment]
[Check this example APKBUILD] (
https://arcetera.moe/git/alpine-ports/file/stagit/APKBUILD.html)

**Comment by dylanaraps** [comment]
How do I install `git` and `bash` in alpine linux? I'm in class so I can't spent time learning how. :)

**Comment by tudurom** [comment]
I think it's

`apk add bash git`

**Comment by dylanaraps** [comment]
They're not found in my virtual machine.

**Comment by dylanaraps** [comment]
I'll figure it out when I get home today.

**Comment by dawidd6** [comment]
''apk update''
''apk add bash git''

Should work, then you need to manually edit ''/etc/passwd'' to change default shell for user

**Comment by dylanaraps** [comment]
Awesome, thanks.

I figured out why the commands weren't working for me, my VM has no internet connection on my laptop.

**Comment by dylanaraps** [comment]
Support for `Alpine Linux` is now in master. @dawidd6

**Comment by dawidd6** [comment]
Great, thanks man :)

**Comment by mclado** [comment]
Hey there! Long time no talk!

Just wanted to share with you that `make install` works flawlessly in Kali linux too, so you can add it under supported selection!

[[http://i.imgur.com/zCgqrrm.png]]

**Comment by iandrewt** [comment]
Multi-battery support can be ignored for OS X, it doesn't support multiple batteries

**Comment by dylanaraps** [comment]
Hey @mclado!

You can embed an image using this syntax:

```sh
![placeholder text while image loads](image url)

# example
![neofetch](imgur.com/fgrrew.png)
```

**Comment by tudurom** [comment]
@ix mind adding support for OpenBSD?

**Comment by dylanaraps** [comment]
## I'll be on vacation from tomorrow till Sunday so don't expect much from me for a few days.

**Comment by bune1991** [comment]
hello thx for your work  can u add image for  OS: KDE neon on 16.04 x86_64
because no image with neon project

**Comment by dylanaraps** [comment]
@bune1991, I don't understand what you're asking. 

- Do you mean ascii art for the neon project?
- Can you send me a screenshot?

**Comment by tudurom** [comment]
I think he asks for a KDE Neon distro ascii art.

**Comment by dylanaraps** [comment]
Ah, I though it was something along those lines. If @bune1991 opens up an issue and provides me with the verbose log, I can easily add an ascii logo. Otherwise I'll have to download KDE Neon overnight and run it in a VM.

**Comment by dylanaraps** [comment]
Sorry guys, I'm very busy atm. 

I'm a week behind with school so I can't dedicate any time to neofetch unfortunately.

**Comment by dylanaraps** [comment]
KDE logo is done, sorry it took so long. 

I've got some time to work on Neofetch again so I'll be pushing commits over the next few days.

**Comment by dylanaraps** [comment]
End of semester is coming up and I'm behind in a few of my classes so I won't be 
working on neofetch very much at all over the next couple of weeks.

I'll still be around to respond to any PRs and support/issues that come up.

Thanks :]

**Comment by dylanaraps** [comment]
Closing this, I'll open a new one soon.
